mobilecoind-dev-faucet/src/data_types.rs

```diff
@@ -188,3 +196,154 @@ impl From<Result<api::SubmitTxResponse, String>> for JsonSubmitTxResponse {
+        }
+        if let Some(val) = req.retries.as_ref() {
+            result.retries = *val;
+        }
+        if let Some(val) = req.retry_period_ms.as_ref() {
+            result.retry_period = Duration::from_millis(*val as u64);
+        }
+        if let Some(val) = req.tombstone_offset.as_ref() {
+            result.tombstone_offset = *val;
+        }
+        if let Some(uris) = req.consensus_uris.as_ref() {
+            result.consensus_client_uris = uris
+                .iter()
+                .map(|uri| FromStr::from_str(uri.as_str()))
+                .collect::<Result<Vec<_>, _>>()
+                .map_err(|err| format!("Invalid uri: {}", err))?;
+        }
+        Ok(result)
+    }
+}
+
+/// A slam resposne includes the parameters used to start the slam, and the
+/// report at the end
+#[derive(Debug, Default)]
+pub struct SlamResponse {
+    /// The slam params actually used
+    pub params: SlamParams,
+    /// The report for the slam operation
+    pub report: SlamReport,
+}
+
+/// Json form of slam params
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub struct JsonSlamParams {
+    /// The target number of txs to submit
+    pub target_num_tx: u32,
+    /// The number of threads to use to submit txs in parallel
+    pub num_threads: u32,
+    /// The number of retries when submitting a transaction
+    pub retries: u32,
+    /// How long to wait before retrying, in milliseconds
+    pub retry_period: u32,
+    /// How many blocks before the tombstone
+    pub tombstone_offset: u32,
+    /// Consensus URIs
+    pub consensus_client_uris: Vec<String>,
+}
+
+impl From<SlamParams> for JsonSlamParams {
+    fn from(src: SlamParams) -> JsonSlamParams {
+        Self {
+            target_num_tx: src.target_num_tx,
+            num_threads: src.num_threads,
+            retries: src.retries,
+            retry_period: src.retry_period.as_millis().try_into().unwrap_or(u32::MAX),
+            tombstone_offset: src.tombstone_offset,
+            consensus_client_uris: src
+                .consensus_client_uris
+                .into_iter()
+                .map(|x| x.to_string())
+                .collect(),
+        }
+    }
+}
+
+/// Json form of a slam report
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub struct JsonSlamReport {
+    /// Num utxos prepared
+    pub num_prepared_utxos: u32,
+    /// Num txs submitted
+    pub num_submitted_txs: u32,
+    /// Prepare duration in milliseconds
+    pub prepare_time: u32,
+    /// Submit duration in milliseconds
+    pub submit_time: u32,
+}
+
+impl From<SlamReport> for JsonSlamReport {
+    fn from(src: SlamReport) -> JsonSlamReport {
+        Self {
+            num_prepared_utxos: src.num_prepared_utxos,
+            num_submitted_txs: src.num_submitted_txs,
+            prepare_time: src.prepare_time.as_millis().try_into().unwrap_or(u32::MAX),
+            submit_time: src.submit_time.as_millis().try_into().unwrap_or(u32::MAX),
```

Comment:
so the reason I had `u32::MAX` was, the `try_into` here is converting `u128` to `u32`, and the error an overflow error. So I just thought saturating at `u32::MAX` was reasonable, since I don't really expect this to happen, it would mean we spent a 4 million seconds at one of these steps. I'm okay to change it to zero in that case I guess